File: backend/src/features/notes/services.py
from datetime import datetime

from fastapi import Depends

from src.core.database import AsyncSession, get_db
from src.features.notes.repository import NoteRepository

from .dto import NoteCreate, NoteUpdate
from .interfaces import INoteRepository


class NoteService:
    def __init__(self, repo: INoteRepository):
        self.repo = repo

    async def create_note(self, note_data: NoteCreate):
        note = await self.repo.create(note_data)
        return note

    async def get_notes(self, day: datetime = None):
        return await self.repo.list_all(day)

    async def update_note(self, note_id: int, note_data: NoteUpdate):
        return await self.repo.update(note_id, note_data)

    async def delete_note(self, note_id: int):
        return await self.repo.delete(note_id)


def get_note_service(db: AsyncSession = Depends(get_db)):
    note_repository = NoteRepository(db)
    return NoteService(note_repository)

File: frontend/src/api/notes/services.ts
import { NoteI } from '@/features/notes/types';
import { NoteRequest } from '@/api/notes/types';
import axios from '@/api/axios';

const notesServices = {
  async getNotes(day: string): Promise<NoteI[]> {
    try {
      const response = await axios.get(`/notes?day=${day}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  async createNote(note: NoteRequest): Promise<NoteI> {
    try {
      const response = await axios.post('/notes', note);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  async deleteNote(id: number): Promise<void> {
    try {
      await axios.delete(`/notes/${id}`);
    } catch (error) {
      throw error;
    }
  },

  async updateNote(id: number, note: NoteRequest): Promise<NoteI> {
    try {
      const response = await axios.put(`/notes/${id}`, note);
      return response.data;
    } catch (error) {
      throw error;
    }
  },
};

export { notesServices };
